sdk/inc/azure/iot/az_iot_common.h

```diff
@@ -61,17 +61,112 @@ typedef enum
+/**
+ * @brief Supported IoT message properties
+ */
+#define AZ_IOT_MESSAGE_PROPERTIES_MESSAGE_ID \
   "%24.mid" /**< Add unique identification to a message */
-#define AZ_IOT_HUB_MESSAGE_PROPERTIES_CORRELATION_ID \
+#define AZ_IOT_MESSAGE_PROPERTIES_CORRELATION_ID \
   "%24.cid" /**< Used in distributed tracing. More information here: \
 https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-distributed-tracing */
-#define AZ_IOT_HUB_MESSAGE_PROPERTIES_CONTENT_TYPE \
+#define AZ_IOT_MESSAGE_PROPERTIES_CONTENT_TYPE \
   "%24.ct" /**< URL encoded and of the form text%2Fplain or application%2Fjson, etc */
-#define AZ_IOT_HUB_MESSAGE_PROPERTIES_CONTENT_ENCODING "%24.ce" /**< UTF-8, UTF-16, etc */
+#define AZ_IOT_MESSAGE_PROPERTIES_CONTENT_ENCODING "%24.ce" /**< UTF-8, UTF-16, etc */
+
+/**
+ * @brief Telemetry or C2D properties.
+ *
+ */
+typedef struct
+{
+  struct
+  {
+    az_span properties_buffer;
+    int32_t properties_written;
+    uint32_t current_property_index;
+  } _internal;
+} az_iot_message_properties;
+
+/**
+ * @brief Initializes the Telemetry or C2D properties.
+ *
+ * @note The properties init API will not encode properties. In order to support
+ *       the following characters, they must be percent-encoded (RFC3986) as follows:
+ *          `/` : `%2F`
+ *          `%` : `%25`
+ *          `#` : `%23`
+ *          `&` : `%26`
+ *       Only these characters would have to be encoded. If you would like to avoid the need to
+ *       encode the names/values, avoid using these characters in names and values.
+ *
+ * @param[in] properties The #az_iot_message_properties to initialize
+ * @param[in] buffer Can either be an empty #az_span or an #az_span containing properly formatted
+ *                   (with above mentioned characters encoded if applicable) properties with the
+ *                   following format: {key}={value}&{key}={value}.
+ * @param[in] written_length The length of the properly formatted properties already initialized
+ * within the buffer. If the \p buffer is empty (uninitialized), this should be 0.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_message_properties_init(
+    az_iot_message_properties* properties,
+    az_span buffer,
+    int32_t written_length);
+
+/**
+ * @brief Appends a key-value property to the list of properties.
+ *
+ * @note The properties append API will not encode properties. In order to support
+ *       the following characters, they must be percent-encoded (RFC3986) as follows:
+ *          `/` : `%2F`
+ *          `%` : `%25`
+ *          `#` : `%23`
+ *          `&` : `%26`
+ *       Only these characters would have to be encoded. If you would like to avoid the need to
+ *       encode the names/values, avoid using these characters in names and values.
+ *
+ * @param[in] properties The #az_iot_message_properties to use for this call
+ * @param[in] name The name of the property.
+ * @param[in] value The value of the property.
```

Comment:
Can these be empty or null, or have whitespace before/after. Might be worth providing these details either here or as remarks.